experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+                                           Gate *originalPhi) {
+
+  // At each iteration, we want to use a cofactor that is present in all the
+  // expressions in `expressions`. Since the cofactors are ordered according to
+  // the basic block number, we can say that if a cofactor is present in one
+  // expression, then it must be present in all the others, since they all have
+  // the blocks associated to that cofactor as common dominator.
+  std::string conditionToUse;
+  while (true) {
+    conditionToUse = conditions.front();
+    conditions.erase(conditions.begin());
+    unsigned cofactorUsage =
+        std::count_if(expressions.begin(), expressions.end(),
+                      [&](std::pair<boolean::BoolExpression *, GateInput *> g) {
+                        return g.first->containsMintern(conditionToUse);
+                      });
+
+    // None is using that cofactor, go to the next one
+    if (cofactorUsage == 0)
+      continue;
+
+    // All are using the cofactor, expand the expressions over it
+    if (cofactorUsage == expressions.size())
+      break;
+
+    assert(false && "A cofactor is used only by some expressions");
+  }
+
+  ListExpressionsPerGate conditionsFalseExpressions;
+  ListExpressionsPerGate conditionsTrueExpressions;
+
+  // For each expression
+  for (auto expression : expressions) {
+
+    // Restrict an expression according to a condition ans an input boolean
+    // value and return it
+    auto getRestrictedExpression =
+        [&](bool value) -> boolean::BoolExpression * {
+      auto *expr = expression.first->deepCopy();
+      boolean::restrict(expr, conditionToUse, value);
+      return expr->boolMinimize();
+    };
+
+    // Substitute a condition in the expression both with value true and false
+    auto *exprTrue = getRestrictedExpression(true);
+    auto *exprFalse = getRestrictedExpression(false);
+
+    // One of the two expressions above will be zero: add the input to the
+    // corresponding list
+    if (exprTrue->toString() != "0")
+      conditionsTrueExpressions.emplace_back(exprTrue, expression.second);
+    if (exprFalse->toString() != "0")
+      conditionsFalseExpressions.emplace_back(exprFalse, expression.second);
+  }
+
+  SmallVector<GateInput *> operandsGamma(2);
+
+  // If the number of non-null expressions obtained with cofactor = X is greater
+  // than 1, then the expansions must be done again over those inputs with a new
+  // cofactor. The resulting gamma is going to be used as X input of this
+  // gamma.
+  //
+  // If only one element is present, that same element is used as input of the
+  // gamma.
+  //
+  // If no elements are present, then that input will never be used, and it is
+  // considered as empty.
+  auto setGammaOperand = [&](int input, ListExpressionsPerGate &list) -> void {
+    if (list.size() > 1) {
+      auto *gamma = expandGammaTree(list, conditions, originalPhi);
+      operandsGamma[input] = new GateInput(gamma);
```

Comment:
Again, this needs to be freed somewhere. Please really pay attention to this, I don't understand how it is possible to use `new` a bunch and never think of using `delete`...

Reply:
yeah that was embarrassing....